akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java

```diff
@@ -0,0 +1,338 @@
+      // instead of minimum, and accuracy will suffer somewhat.
+      // Probably makes sense to add an "allow negative increments"
+      // parameter to constructor.
+      throw new IllegalArgumentException("Negative increments not implemented");
+    }
+    final int[] buckets = MurmurHash.hashBuckets(item, depth, width);
+    for (int i = 0; i < depth; ++i) {
+      table[i][buckets[i]] += count;
+    }
+    size += count;
+  }
+
+  /**
+   * Similar to {@code add}, however we reuse the fact that the hask buckets have to be calculated for {@code add}
+   * already, and a separate {@code estimateCount} operation would have to calculate them again, so we do it all in one go.
+   */
+  public long addAndEstimateCount(String item, long count) {
+    if (count < 0) {
+      // Actually for negative increments we'll need to use the median
+      // instead of minimum, and accuracy will suffer somewhat.
+      // Probably makes sense to add an "allow negative increments"
+      // parameter to constructor.
+      throw new IllegalArgumentException("Negative increments not implemented");
+    }
+    final int[] buckets = MurmurHash.hashBuckets(item, depth, width);
+    for (int i = 0; i < depth; ++i) {
+      table[i][buckets[i]] += count;
+    }
+    size += count;
+    return estimateCount(buckets);
+  }
+
+  public long size() {
+    return size;
+  }
+  
+  /**
+   * The estimate is correct within {@code 'epsilon' * (total item count)},
+   * with probability {@code confidence}.
+   */
+  public long estimateCount(long item) {
+    long res = Long.MAX_VALUE;
+    for (int i = 0; i < depth; ++i) {
+      res = Math.min(res, table[i][hash(item, i)]);
+    }
+    return res;
+  }
+
+  /**
+   * The estimate is correct within {@code 'epsilon' * (total item count)},
+   * with probability {@code confidence}.
+   */
+  public long estimateCount(String item) {
+    int[] buckets = MurmurHash.hashBuckets(item, depth, width);
+    return estimateCount(buckets);
+  }
+
+  /**
+   * The estimate is correct within {@code 'epsilon' * (total item count)},
+   * with probability {@code confidence}.
+   * 
+   * @param buckets the "indexes" of buckets from which we want to calculate the count
+   */
+  private long estimateCount(int[] buckets) {
+    long res = Long.MAX_VALUE;
+    for (int i = 0; i < depth; ++i) {
+      res = Math.min(res, table[i][buckets[i]]);
+    }
+    return res;
+  }
+  
+  /**
+   * This is copied from https://github.com/addthis/stream-lib/blob/master/src/main/java/com/clearspring/analytics/hash/MurmurHash.java
+   * Which is Apache V2 licensed.
+   * <p></p>
+   * This is a very fast, non-cryptographic hash suitable for general hash-based
+   * lookup. See http://murmurhash.googlepages.com/ for more details.
+   * <p/>
+   * <p>
+   * The C version of MurmurHash 2.0 found at that site was ported to Java by
+   * Andrzej Bialecki (ab at getopt org).
+   * </p>
+   */
+  private static class MurmurHash {
```

Comment:
I did some benchmark of hashes:

```
[info] Benchmark               (lengthOfArray)   Mode  Cnt   Score   Error   Units
[info] HashesBench.adler32                 100  thrpt    6  14,354 ? 2,779  ops/us
[info] HashesBench.adler32                 200  thrpt    6   9,207 ? 0,312  ops/us
[info] HashesBench.crc32                   100  thrpt    6  11,483 ? 6,097  ops/us
[info] HashesBench.crc32                   200  thrpt    6   9,686 ? 0,190  ops/us
[info] HashesBench.fnv32                   100  thrpt    6   9,871 ? 0,011  ops/us
[info] HashesBench.fnv32                   200  thrpt    6   4,533 ? 0,006  ops/us
[info] HashesBench.fnv64                   100  thrpt    6   9,838 ? 0,033  ops/us
[info] HashesBench.fnv64                   200  thrpt    6   4,519 ? 0,018  ops/us
[info] HashesBench.murmur3                 100  thrpt    6  13,745 ? 0,114  ops/us
[info] HashesBench.murmur3                 200  thrpt    6   7,707 ? 1,650  ops/us
[info] HashesBench.murmurAkka              100  thrpt    6   3,531 ? 0,044  ops/us
[info] HashesBench.murmurAkka              200  thrpt    6   1,570 ? 0,237  ops/us
```

Previous results were rather wonky (some hashes said 800k ops/us wat) so I stored the result in a volatile var in the benchmark. These results look better. Murmur3 is `scala.util.hashing.MurmurHash3`, FNV are implementations from the internet, murmurAkka is the one that is included in akka itself. CRC32 and Adler32 are the JDK provided ones.